### DISCUSSION TITLE
Filling ErrorWithDiff fileds e.g. expected, received, diff from custom Assertion code

### DISCUSSION THREAD

**Discussion by jkavalec** [vitest-dev/vitest#5829] [Q&A] [answered]
Hi, 

I am in refactoring phase of my Assertion Lib + Reporter Lib, and one of last things I need to clean up is unifying handling for my Assertion Lib, with Vitest default Assertions.

When within Reporter you process the data from Assertion error within ErrorWithDiff object. Vitest assertions have fields like expected, diff etc. filled in, pseudocode bellow demonstrates core of my assertion logic, from this I get only stackStr filled in automatically, other fileds are unsuprisingly empty. I was not able to find out within some light research how could these fields be filled by me manually, i.e. logic within Vitest code that does this filling.

```
export const match = (received, expected) => {
  const error = new Error(generate_error_msg(received, expected))
  Error.captureStackTrace(error, match)
  throw error
}
```

I would appreciate any leads to help me solve this, so far most things work well, but that part of parsing stackStr is pretty messy plus, I have some cases that would make use of these values being accessible in ErrorWithDiff object.

**Comment by sheremet-va** [comment, marked as answer]
The diff is generated if the thrown error already has `expected` and `actual` fields:

```ts
export const match = (received, expected) => {
  const error = new Error(generate_error_msg(received, expected))
  Error.captureStackTrace(error, match)
  error.actual = received
  error.expected = expected
  throw error
}
```

The error is processed by this function in the test thread:

https://github.com/vitest-dev/vitest/blob/70a277b5abc8b5586542adec8de10edb43a3e96f/packages/utils/src/error.ts#L97

**Reply by jkavalec** [reply]
Great, thank you for clarifying this, works as I wanted it to, I was skimming through that function but failed to read the filling code properly.
